### PR TITLE
[GEARPUMP-350] Fix the not started app clock

### PR DESCRIPTION
**Pull request by huafengw**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the commit message is formatted like:
   `[GEARPUMP-<Jira issue #>] Meaningful description of pull request` 
 - [ ] Make sure tests pass via `sbt clean test`.
 - [ ] Make sure old documentation affected by the pull request has been updated and new documentation added for new functionality.

### CODE REVIEW

**Comment by huafengw** [comment]
Verified streaming examples including complex dag, SOL, word count dsl and wordcount java.

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala [hunk 1]
```diff
@@ -254,6 +254,7 @@ class TaskActor(
           case m: Message =>
             count += 1
             onNext(m)
+            processingWatermark = TaskUtil.max(processingWatermark, m.timestamp)
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: This doesn't look right for window processing watermark which **can not** advance on every new message

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/WatermarkProducer.scala [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.task
+
+trait WatermarkProducer
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/WatermarkProducer.scala]: Please add comments on the purpose of this interface

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskUtil.scala [hunk 3]
```diff
@@ -53,6 +53,10 @@ object TaskUtil {
     else t1
   }
 
+  def min(t1: Instant, t2: Instant, t3: Instant): Instant = {
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskUtil.scala]: If this is a util, then how about making it more general by accepting *varargs* ?

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala [hunk 4]
```diff
@@ -361,10 +358,11 @@ class TaskActor(
   }
 
   private def getSubscriptionWatermark(subs: List[(Int, Subscription)], wmk: Instant): Instant = {
+    val wmkInMilli = wmk.toEpochMilli
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: what is this change for ?

File: streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala [hunk 5]
```diff
@@ -27,15 +27,15 @@ import org.apache.gearpump.cluster.UserConfig
...
+import org.apache.gearpump.streaming.task.{Task, TaskContext, TaskUtil, WatermarkProducer}
 
 /**
  * Processes messages in groups as defined by groupBy function.
  */
 class GroupByTask[IN, GROUP, OUT](
     groupBy: IN => GROUP,
     taskContext: TaskContext,
-    userConfig: UserConfig) extends Task(taskContext, userConfig) {
+    userConfig: UserConfig) extends Task(taskContext, userConfig) with WatermarkProducer {
```
Comment (manuzhang) [inline streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala]: Why is `GroupByTask` a `WatermarkProducer` ?

Comment (manuzhang) [reply streaming/src/main/scala/org/apache/gearpump/streaming/dsl/task/GroupByTask.scala]: And only `GroupByTask` ?

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala (hunk 4 above)
Comment (huafengw) [reply streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: To avoid multiple `toEpochMilli` calls

**Comment by manuzhang** [comment]
@huafengw This can be closed now

**Pull request closed without merging**